### PR TITLE
Puppeteer page error - Fix bundle

### PR DESCRIPTION
**Pull request by hummal**
This PR is trying to fix 2 main problems:
- `Cannot find context with specified id undefined`
- `Execution context was destroyed. undefined`

Additional it is taking care of some other minor issues:
✅ added missing dependecies for development
✅ added `.gitattributes` file to ensure correct line endings and file handlings on all os
✅ fixed `package.json` target nodejs version. It has to be string instead of float
✅ moved the `pageLoadSkipTimeout `calculation into the browser **after** the first response (_document_) to ensure a better timing and remove redundancy with timeout
✅ `pageLoadSkipTimeout `now prevents the page to load additional resources after timing out with `window.stop()`
✅ fixed error handling and promise rejecting. Now all pages which throw errors will repeated once like they should and dont break penthouse
✅ handling of https://github.com/GoogleChrome/puppeteer/issues/1591 bug with` page.waitFor(n)`
✅ slightly reformating

### CODE REVIEW

**Comment by hummal** [comment]
Accoding the failing build: 
https://github.com/babel/babel/tree/master/packages/babel-preset-env

> Similarly, if you're targeting Node.js instead of the browser, you can configure @babel/preset-env to only include the polyfills and transforms necessary for a particular version:
> 
> ```
> {
>   "presets": [
>     ["@babel/preset-env", {
>       "targets": {
>         "node": "6.10"
>       }
>     }]
>   ]
> }
> ```

**Comment by hummal** [comment]
by the way is there a reason why you dont update your dependecies? 
```
 jsesc                    ^1.0.0  →          ^2.5.1
 puppeteer                  0.12  →             1.0
 chai                     ^1.9.1  →          ^4.1.2
 eslint-config-standard  ^10.2.1  →  ^11.0.0-beta.0
 eslint-plugin-node       ^5.2.1  →          ^6.0.0
 husky                   ^0.13.3  →         ^0.14.3
 lint-staged              ^3.4.1  →          ^6.1.0
 mocha                   ^1.20.1  →          ^5.0.0
 prettier-standard        ^5.0.0  →          ^8.0.0

The following dependencies are satisfied by their declared version range, but the installed versions are behind. You can install the latest versions without modifying your package file by using npm update.
 If you want to update the dependencies in your package file anyway, run ncu -a.

 jest      ^22.1.4  →  ^22.2.0
 eslint    ^4.11.0  →  ^4.17.0
 gm        ^1.21.1  →  ^1.23.1
 prettier   ^1.3.1  →  ^1.10.2
```

**Comment by hummal** [comment]
I updated the dependecies to the latest versions (except puppeteer). Removed mocha and chai (because you are using jest) and it worked till the extra and perf tests. It has somehow a problem with creating the penthouse instance.

You should propably check that. I won't commit these changes right now but you can see the results in here: https://travis-ci.org/hummal/penthouse

**Comment by pocketjoso** [comment]
So thanks for the PR, I started reviewing now. There's some stuff in here that is just obvious improvement, but then there are other things that we need to talk about before merging.
I looked at whether we could cherry-pick out the easy changes from your commits, but they were spread out across commits, so [I opened up a separate PR with just those changes](https://github.com/pocketjoso/penthouse/pull/228). If you are up for it, please re-open your own PR with just those changes, so the commits get credited to you, since the changes are yours. Otherwise I will merge that, and we will just continue with the other changes here.

File: src/browser-sandbox/pruneNonCriticalSelectors.js [hunk 1]
```diff
@@ -83,19 +83,25 @@ export default function pruneNonCriticalSelectors ({
...
 
-  // give some time (renderWaitTime) for sites like facebook that build their page dynamically,
-  // otherwise we can miss some selectors (and therefor rules)
-  // --tradeoff here: if site is too slow with dynamic content,
-  // it doesn't deserve to be in critical path.
-  return sleep(renderWaitTime).then(() => {
-    console.log('debug: waited for renderWaitTime: ' + renderWaitTime)
-    return filterSelectors(selectors)
-  })
+  try {
```
Comment (pocketjoso) [inline src/browser-sandbox/pruneNonCriticalSelectors.js]: Did you test putting this try/catch inside core.js instead?  As long as that still catches the error, I prefer it, as we can react to the error, instead of just receiving no selectors from this function. I implemented this in [my branch here](https://github.com/pocketjoso/penthouse/pull/228/files#diff-7eb52b366866677666470e019283c8eaR177), but I don't have an easy way to test the issues you have been testing, so I don't know 100% that it catches your errors, but it looks like it should.

File: src/core.js [hunk 2]
```diff
@@ -101,52 +124,105 @@ async function pruneNonCriticalCssLauncher ({
...
-                    's. '
-                )
-                resolve()
-              }
-            }, pageLoadSkipTimeout)
+      const loadPageResponse = new Promise((resolve, reject) => {
+        page
+          .goto(url, {
+            timeout: timeout + 1000,
+            waitUntil: 'networkidle'
```
Comment (pocketjoso) [inline src/core.js]: Hmm, I tried this value before but it leads some urls to get stuck, if they continuously send requests for tracking and similar. I'm not sure it is a better default for Penthouse. What was your reasoning behind this change?

File: src/core.js [hunk 3]
```diff
@@ -101,52 +124,105 @@ async function pruneNonCriticalCssLauncher ({
...
+          .catch(err => {
+            _hasError = err
+            // Reject when error because we don't want an errored page
+            console.error(err)
+            return reject(err)
+          })
+      })
+
+      try {
+        // Instead of race we want to use all for better workflow. Race is evil
```
Comment (pocketjoso) [inline src/core.js]: this comment doesn't match the code, did you change your mind or is it a mistake? I am interested in why you think 'Race is evil'? To the best of my knowledge so far, Promise.race is a good fit here.

File: src/index.js [hunk 4]
```diff
@@ -52,7 +53,8 @@ const launchBrowserIfNeeded = async function ({ getBrowser }) {
         // for spammy pages that keep on sending (non critcal) requests
         waitUntil: 'networkidle2',
         ignoreHTTPSErrors: true,
-        args: ['--disable-setuid-sandbox', '--no-sandbox']
+        args: ['--disable-setuid-sandbox', '--no-sandbox'],
+        dumpio: false
```
Comment (pocketjoso) [inline src/index.js]: This seems to be the default in the [latest version of Puppeteer](https://github.com/GoogleChrome/puppeteer/blob/master/docs/api.md#puppeteerlaunchoptions). Why did you add it; is it not the default on the 0.12 version?

File: src/core.js [hunk 5]
```diff
@@ -166,11 +242,11 @@ async function pruneNonCriticalCssLauncher ({
       )
       debuglog('build selector profile')
 
+      await page.waitFor(500)
```
Comment (pocketjoso) [inline src/core.js]: this is the kind of change that I most like would want as an opt-in, or even just configurable so that only users who need to change this can do so, without slowing down penthouse for everyone.

File: src/browser-sandbox/pageLoadSkipTimeout.js [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+  try {
+    return new Promise((resolve, reject) => {
+      if (pageLoadSkipTimeout) {
+        const start = Date.now()
+        const pageLoadTimeout = function () {
+          window.requestAnimationFrame(() => {
+            const timePassed = Date.now() - start
+
+            if (timePassed >= pageLoadSkipTimeout) {
+              window.stop()
```
Comment (pocketjoso) [inline src/browser-sandbox/pageLoadSkipTimeout.js]: The intention with the Promise.race for the page load in `master` is just to prevent Penthouse to take too long, when the page has reasonably loaded, but due to imperfect detection mechanisms puppeteer still believes it has not.

File: src/core.js [hunk 7]
```diff
@@ -101,52 +124,105 @@ async function pruneNonCriticalCssLauncher ({
...
+
+      // Handle requests of pages and block the page from going away and losing context
+      // https://github.com/pocketjoso/penthouse/issues/202
+      let pageLoadSkipPromise = new Promise((resolve, reject) => {
+        if (pageLoadSkipTimeout) {
+          // enables us to control requests(stop, continue, modify)
+          page.on('response', async response => {
+            if (response.url === url) {
+              debuglog('RESPONSE URL: ' + response.url) // TODO: interceptedRequest.url is a function in puppeteer > 0.12
+              if (pageLoadSkipTimeout) {
```
Comment (pocketjoso) [inline src/core.js]: you already checked this

File: src/core.js [hunk 8]
```diff
@@ -101,52 +124,105 @@ async function pruneNonCriticalCssLauncher ({
...
-              }
-            }, pageLoadSkipTimeout)
+      const loadPageResponse = new Promise((resolve, reject) => {
+        page
+          .goto(url, {
+            timeout: timeout + 1000,
+            waitUntil: 'networkidle'
+          })
+          .then(response => {
+            // Reject exits the Promise.all call immediately and ensures that the pageload is always higher prio than timout
```
Comment (pocketjoso) [inline src/core.js]: I think you are mixing up Promise.all and Promise.race here - this comment is not accurate. (this fulfills the promise, hopefully winning the race. Would not do what you want if it was promise.all below).

File: package.json [hunk 9]
```diff
@@ -83,7 +85,7 @@
         "env",
         {
           "targets": {
-            "node": 6.4
+            "node": "6.4"
```
Comment (pocketjoso) [inline package.json]: I remember thinking so to, but see the errors  this gives:

File: src/browser-sandbox/pruneNonCriticalSelectors.js (hunk 1 above)
Comment (hummal) [reply src/browser-sandbox/pruneNonCriticalSelectors.js]: The reason for me to do it that way was to ensure to solve the execution context problem. But I tried a lot of things. So I will check if its running safe with the try catch in core. In common you are right.

File: src/core.js (hunk 2 above)
Comment (hummal) [reply src/core.js]: This is supposed to your other questions. 
I did the window.stop() in the executed page tab because I want the load event get fired. In this case the timeout wins the race with a correct "window load event". I did some testing with that but after all the changes I made I wouldnt bet it wouldn't run without it.

File: src/core.js (hunk 3 above)
Comment (hummal) [reply src/core.js]: The comment is just wrong. I changed it to Promise.all to be more safe. To explain:
The evil thing on race is, if you reject a promise the race is over and the promise race is rejected at all. But what we want to achieve is that the first promise resolve will win. Nevertheless I needed to change the workflow of promise.all to be reversed (reject is win). So I decided it is also not the best way. I just changed my mind. yeah

File: src/index.js (hunk 4 above)
Comment (hummal) [reply src/index.js]: I missed to remove this. Just used it to dump all console output while searching for the execution context problem. Can be removed

File: src/core.js (hunk 5 above)
Comment (hummal) [reply src/core.js]: Maybe you are right. I need to use it in a very large scale and Iam in need of a working system. If there is a chance of an error my results will be wrong and this is crucial and could cost thousands of Euros.
But of course, we could make an options out of it. Good idea

File: src/browser-sandbox/pageLoadSkipTimeout.js (hunk 6 above)
Comment (hummal) [reply src/browser-sandbox/pageLoadSkipTimeout.js]: The mindset of stop waiting for the page to load but let it load while extracting the css is helping me to understand. 
As mentioned above I guess it is triggering the "load" event which results in a resolving of the promise. Which is what I want to achieve.

File: src/core.js (hunk 7 above)
Comment (hummal) [reply src/core.js]: you are right

File: src/core.js (hunk 8 above)
Comment (hummal) [reply src/core.js]: It is a relic of my try to replace race with all. Just a wrong comment.

File: package.json (hunk 9 above)
Comment (hummal) [reply package.json]: I know that travis is dealing so. But I really dont get why. This is more a travis problem then a bug. Because as mentioned in the documenation it is recommended to use a string. 
What do you want to do now?

**Comment by hummal** [comment]
At all, Iam for vacation till wednesday. I can't pull a new pr until then. If you can't wait, you should go.

**Comment by pocketjoso** [comment]
Okay, so I think what I'm going to do:
1. merge #224 - which I realised after looking at it again not only improves performance of penthouse further (5-10% faster execution time on average), but also cleans up the code a bit.
2. then I will rebase #228 with your ease changes and get that merged as well.
(both of these before Wednesday)

After that we would need to rebase this PR to take up the changes - I would however like you @hummal to give what I wrote in https://github.com/pocketjoso/penthouse/issues/202#issuecomment-364698843 a go first.

And finally, please open a separate PR for your branch with dependency bumps (https://travis-ci.org/hummal/penthouse) - as long as the tests pass (I disabled a performance test that was failing before) we can get that merged quickly too.

Cheers, keep up the good work! 👏

File: src/browser-sandbox/pruneNonCriticalSelectors.js (hunk 1 above)
Comment (pocketjoso) [reply src/browser-sandbox/pruneNonCriticalSelectors.js]: tested and merged this in (via core.js) in #228

File: src/core.js (hunk 2 above)
Comment (pocketjoso) [reply src/core.js]: I actually see that Penthouse uses `networkidle2`, but specified in the puppeteer.launch call [here](https://github.com/pocketjoso/penthouse/blob/master/src/index.js#L51)... which according to the existing documentation seems wrong - although they might have moved it in a more recent version.
So all in all, I would be in favour of testing the `networkidle2` value here - but unless it is related to your changes I suggest for you _not_ to do it (in this PR), and I can isolate and test just that change myself (later).

**Comment by pocketjoso** [comment]
So, an update from me:
I did merge #224, rebase the simplest changes you had and merge those as well via #228, and I am currently sanity testing these changes under `penthouse@1.4.1-rc3` - only expecting slightly faster execution on average.

So what remains here? For me the next step would be for you @hummal to try https://github.com/pocketjoso/penthouse/issues/202#issuecomment-364698843 as I said above (ideally on updated master), then if you still have problems, let's continue with the solutions for them in this PR, which would then need to be rebased.

**Comment by hummal** [comment]
Ah I see my comment wasn't submitted before. 
I tried to use puppeteer as browser before, but I can't get it to work. I always run into the error "ERROR: kill ERSCH" and I can't fix it. I remembered that I tried that before. 

As for all the pr requests. I don't have so many time that I want to use it to have my name in a commit I did. If you want to use this extra pr to split the commits, it is ok for me if you just merge yours with my changes. We both know that I supported you. And the software is better, so WIN-WIN.

As I said, Iam not able to test your suggestions of https://github.com/pocketjoso/penthouse/issues/202#issuecomment-364698843 

What else is left?

**Comment by hummal** [comment]
Merged current master into my pr

**Comment by hummal** [comment]
Split into other PRs.
